File: python/algorithm/sorting/quick.py
def sort(nums):
  return quickSort(nums, 0, len(nums) - 1)

def quickSort(nums, pivot, end):
  store = pivot + 1
  for i in range (store, end + 1):
    if nums[pivot] > nums[i]:
      swap(nums, i, store)
      store += 1
  swap(nums, pivot, store - 1)

  if pivot < store - 2:
    quickSort(nums, pivot, store - 2)
  if store < end:
    quickSort(nums, store, end)

  return nums

def swap(nums, i, j):
  tmp = nums[i]
  nums[i] = nums[j]
  nums[j] = tmp

File: ts/algorithm/sorting/quick.ts
export function Quick (nums: Array<number>): Array<number> {
  return sort(nums, 0, nums.length - 1)
}

function sort(nums: Array<number>, pivotIdx: number, endIdx: number): Array<number> {
  let storeIdx = pivotIdx + 1
  for (let i = storeIdx; i <= endIdx; i++) {
    if (nums[i] < nums[pivotIdx]) {
      swap(nums, i, storeIdx)
      storeIdx++
    }
  }
  swap(nums, pivotIdx, storeIdx - 1)

  if (pivotIdx < storeIdx - 2)
    sort(nums, pivotIdx, storeIdx -2)
  if (storeIdx < endIdx)
    sort(nums, storeIdx, endIdx)

  return nums
}

function swap (nums: Array<number>, i: number, j: number) {
  let tmp = nums[i]
  nums[i] = nums[j]
  nums[j] = tmp
}
